Can't fix variable when bounds are defined.

If a group of variables has been defined within certain bounds, then it is not possible to fix the value of one of them. In many applications, it is valuable to create a group of variables over a set and then be able to fix only of them.

This code will reproduce the error on Master

```Julia
m = Model() 
@variable(m, 0.0 <= x[1:2] <= 10.0)
JuMP.fix(x[1],0.0) 


AssertionError: !(has_upper_bound(v)) && !(has_lower_bound(v))

Stacktrace:
 [1] fix(::VariableRef, ::Float64) at /Users/jdlara/.julia/packages/JuMP/aU7VH/src/variables.jl:407
 [2] top-level scope at In[50]:3
```

--- a/docs/src/variables.md
+++ b/docs/src/variables.md
@@ -189,7 +189,7 @@ julia> JuMP.lower_bound(x)
 2.0
 ```
 
-Finally, we can delete variable bounds using `JuMP.delete_lower_bound` and
+We can delete variable bounds using `JuMP.delete_lower_bound` and
 `JuMP.delete_upper_bound`:
 ```jldoctest; setup=:(model=Model())
 julia> @variable(model, 1 <= x <= 2)
@@ -210,6 +210,40 @@ julia> JuMP.delete_upper_bound(x)
 
 julia> JuMP.has_upper_bound(x)
 false
+```
+
+In addition to upper and lower bounds, JuMP variables can also be fixed to a
+value.
+```jldoctest; setup=:(model=Model())
+julia> @variable(model, x == 1)
+x
+
+julia> JuMP.is_fixed(x)
+true
+
+julia> JuMP.fix_value(x)
+1.0
+
+julia> JuMP.unfix(x)
+
+julia> JuMP.is_fixed(x)
+false
+```
+Fixing a variable with existing bounds will throw an error. To delete the bounds
+prior to fixing, use `JuMP.fix(variable, value; force = true)`.
+
+```jldoctest; setup=:(model=Model())
+julia> @variable(model, x >= 1)
+x
+
+julia> JuMP.fix(x, 2)
+ERROR: Unable to fix x to 2 because it has existing variable bounds. Consider calling `JuMP.fix(variable, value; force=true)` instead.
+
+julia> JuMP.fix(x, 2; force = true)
+
+
+julia> JuMP.fix_value(x)
+2.0
 ```
 
 ## Variable containers

--- a/docs/src/variables.md
+++ b/docs/src/variables.md
@@ -237,7 +237,7 @@ julia> @variable(model, x >= 1)
 x
 
 julia> JuMP.fix(x, 2)
-ERROR: Unable to fix x to 2 because it has existing variable bounds. Consider calling `JuMP.fix(variable, value; force=true)` instead.
+ERROR: Unable to fix x to 2 because it has existing variable bounds. Consider calling `JuMP.fix(variable, value; force=true)` which will delete existing bounds before fixing the variable.
 
 julia> JuMP.fix(x, 2; force = true)
 

--- a/src/variables.jl
+++ b/src/variables.jl
@@ -411,9 +411,8 @@ create a new one.
 
 If the variable already has variable bounds and `force=false`, calling `fix`
 will throw an error. If `force=true`, existing variable bounds will be deleted,
-and the fixing constraint will be added. Note that subsequently calling
-`JuMP.unfix` will not revert the variable to its previous bounds; users are
-expected to re-add these manually.
+and the fixing constraint will be added. Note a variable will have no bounds
+after a call to `JuMP.unfix`.
 """
 function fix(variable::VariableRef, value::Number; force::Bool = false)
     new_set = MOI.EqualTo(convert(Float64, value))

--- a/src/variables.jl
+++ b/src/variables.jl
@@ -411,7 +411,9 @@ create a new one.
 
 If the variable already has variable bounds and `force=false`, calling `fix`
 will throw an error. If `force=true`, existing variable bounds will be deleted,
-and the variable will be fixing constraint will be added.
+and the fixing constraint will be added. Note that subsequently calling
+`JuMP.unfix` will not revert the variable to its previous bounds; users are
+expected to re-add these manually.
 """
 function fix(variable::VariableRef, value::Number; force::Bool = false)
     new_set = MOI.EqualTo(convert(Float64, value))

--- a/src/variables.jl
+++ b/src/variables.jl
@@ -404,23 +404,40 @@ function set_fix_index(v::VariableRef, cindex::MOIFIX)
 end
 
 """
-    fix(v::VariableRef,upper::Number)
+    fix(v::VariableRef, value::Number; force::Bool = false)
 
-Fix a variable to a value. Update the fixing constraint if one exists, otherwise create a new one.
+Fix a variable to a value. Update the fixing constraint if one exists, otherwise
+create a new one.
+
+If the variable already has variable bounds and `force=false`, calling `fix`
+will throw an error. If `force=true`, existing variable bounds will be deleted,
+and the variable will be fixing constraint will be added.
 """
-function fix(v::VariableRef,upper::Number)
-    newset = MOI.EqualTo(convert(Float64,upper))
-    # are we already fixed?
-    if is_fixed(v)
-        cindex = fix_index(v)
-        MOI.set(backend(owner_model(v)), MOI.ConstraintSet(), cindex, newset)
-    else
-        @assert !has_upper_bound(v) && !has_lower_bound(v) # Do we want to remove these instead of throwing an error?
-        cindex = MOI.add_constraint(backend(owner_model(v)),
-                                    MOI.SingleVariable(index(v)), newset)
-        set_fix_index(v, cindex)
+function fix(variable::VariableRef, value::Number; force::Bool = false)
+    new_set = MOI.EqualTo(convert(Float64, value))
+    model = backend(owner_model(variable))
+    if is_fixed(variable)  # Update existing fixing constraint.
+        c_index = fix_index(variable)
+        MOI.set(model, MOI.ConstraintSet(), c_index, new_set)
+    else  # Add a new fixing constraint.
+        if has_upper_bound(variable) || has_lower_bound(variable)
+            if !force
+                error("Unable to fix $(variable) to $(value) because it has " *
+                      "existing variable bounds. Consider calling " *
+                      "`JuMP.fix(variable, value; force=true)` instead.")
+            end
+            if has_upper_bound(variable)
+                delete_upper_bound(variable)
+            end
+            if has_lower_bound(variable)
+                delete_lower_bound(variable)
+            end
+        end
+        c_index = MOI.add_constraint(
+            model, MOI.SingleVariable(index(variable)), new_set)
+        set_fix_index(variable, c_index)
     end
-    nothing
+    return
 end
 
 """

--- a/src/variables.jl
+++ b/src/variables.jl
@@ -424,7 +424,8 @@ function fix(variable::VariableRef, value::Number; force::Bool = false)
             if !force
                 error("Unable to fix $(variable) to $(value) because it has " *
                       "existing variable bounds. Consider calling " *
-                      "`JuMP.fix(variable, value; force=true)` instead.")
+                      "`JuMP.fix(variable, value; force=true)` which will " *
+                      "delete existing bounds before fixing the variable.")
             end
             if has_upper_bound(variable)
                 delete_upper_bound(variable)

--- a/test/JuMPExtension.jl
+++ b/test/JuMPExtension.jl
@@ -118,14 +118,16 @@ function JuMP.delete_upper_bound(vref::MyVariableRef)
 end
 JuMP.is_fixed(vref::MyVariableRef) = variable_info(vref).has_fix
 JuMP.fix_value(vref::MyVariableRef) = variable_info(vref).fixed_value
-function JuMP.fix(vref::MyVariableRef, value)
+function JuMP.fix(vref::MyVariableRef, value; force::Bool = false)
     info = variable_info(vref)
-    update_variable_info(vref,
-                         JuMP.VariableInfo(info.has_lb, info.lower_bound,
-                                           info.has_ub, info.upper_bound,
-                                           true, value,
-                                           info.has_start, info.start,
-                                           info.binary, info.integer))
+    if !force && (info.has_lb || info.has_ub)
+        error("Unable to fix $(vref) to $(value) because it has existing bounds.")
+    end
+    update_variable_info(vref, JuMP.VariableInfo(
+        false, info.lower_bound, false, info.upper_bound, true, value,
+        info.has_start, info.start, info.binary, info.integer)
+    )
+    return
 end
 function JuMP.unfix(vref::MyVariableRef)
     info = variable_info(vref)

--- a/test/variable.jl
+++ b/test/variable.jl
@@ -140,6 +140,13 @@ function test_variable_fix(ModelType)
     @test JuMP.fix_value(fixed) == 1.0
     JuMP.unfix(fixed)
     @test !JuMP.is_fixed(fixed)
+    JuMP.set_lower_bound(fixed, 0.0)
+    @test_throws Exception JuMP.fix(fixed, 1.0)
+    JuMP.fix(fixed, 1.0; force = true)
+    @test !JuMP.has_lower_bound(fixed)
+    @test !JuMP.has_upper_bound(fixed)
+    @test JuMP.is_fixed(fixed)
+    @test JuMP.fix_value(fixed) == 1.0
 end
 
 function test_variable_custom_index_sets(ModelType)

